python/ray/serve/_private/replica.py

```diff
@@ -417,55 +416,52 @@ async def handle_request(
-        via an ASGI send interface.
+        The user method is called in a `Task` and places its results on a
+        `result_queue`. This method pulls and yields from the `result_queue`.
         """
-        receiver_task = None
         call_user_method_task = None
         wait_for_message_task = None
         try:
-            receiver = ASGIReceiveProxy(
-                request_metadata.request_id, request.http_proxy_handle
-            )
-            receiver_task = self._event_loop.create_task(
-                receiver.fetch_until_disconnect()
-            )
-
-            scope = pickle.loads(request.pickled_asgi_scope)
-            asgi_queue_send = ASGIMessageQueue()
-            request_args = (scope, receiver, asgi_queue_send)
-            request_kwargs = {}
-
             # Handle the request in a background asyncio.Task. It's expected that
-            # this task will use the provided ASGI send interface to send its HTTP
-            # the response. We will poll for the sent messages and yield them back
-            # to the caller.
+            # this task will use the result queue to send its response messages.
+            result_queue = MessageQueue()
             call_user_method_task = self._event_loop.create_task(
                 self._user_callable_wrapper.call_user_method(
-                    request_metadata, request_args, request_kwargs
+                    request_metadata,
+                    request_args,
+                    request_kwargs,
+                    generator_result_callback=result_queue,
                 )
             )
 
             while True:
                 wait_for_message_task = self._event_loop.create_task(
-                    asgi_queue_send.wait_for_message()
+                    result_queue.wait_for_message()
                 )
                 done, _ = await asyncio.wait(
                     [call_user_method_task, wait_for_message_task],
                     return_when=asyncio.FIRST_COMPLETED,
                 )
+
                 # Consume and yield all available messages in the queue.
-                # The messages are batched into a list to avoid unnecessary RPCs and
-                # we use vanilla pickle because it's faster than cloudpickle and we
-                # know it's safe for these messages containing primitive types.
-                yield pickle.dumps(asgi_queue_send.get_messages_nowait())
+                messages = result_queue.get_messages_nowait()
+                if messages:
+                    # HTTP (ASGI) messages are only consumed by the proxy so batch them
+                    # and use vanilla pickle (we know it's safe because these messages
+                    # only contain primitive Python types).
+                    if request_metadata.is_http_request:
+                        yield pickle.dumps(messages)
+                    else:
+                        for msg in messages:
+                            yield msg
```

Comment:
Should we `yield pickle.dumps(messages)` for gRPC too? I assume the `else` case is only if the request originated directly through a `ServeHandle`?

Reply:
we can, would require also refactoring the proxy. you're right that the `else` case is mostly for deployment handle calls

Reply:
Wait, why do we need pickling gRPC? We should serialize it to bytes just as Proto

Reply:
yes grpc messages are serialized as bytes already. this would just be opportunistically batching them into a list (the `pickle` part is unnecessary)